nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/main/java/org/apache/nifi/processors/kafka/KafkaPublisher.java

```diff
@@ -0,0 +1,159 @@
+        this.producer = new Producer<>(producerConfig);
+    }
+
+    /**
+     *
+     */
+    void setProcessLog(ProcessorLog processLog) {
+        this.processLog = processLog;
+    }
+
+    /**
+     * Publishes messages to Kafka topic. It supports three publishing
+     * mechanisms.
+     * <ul>
+     * <li>Sending the entire content stream as a single Kafka message.</li>
+     * <li>Splitting the incoming content stream into chunks and sending
+     * individual chunks as separate Kafka messages.</li>
+     * <li>Splitting the incoming content stream into chunks and sending only
+     * the chunks that have failed previously @see
+     * {@link SplittableMessageContext#getFailedSegments()}.</li>
+     * </ul>
+     * This method assumes content stream affinity where it is expected that the
+     * content stream that represents the same Kafka message(s) will remain the
+     * same across possible retries. This is required specifically for cases
+     * where delimiter is used and a single content stream may represent
+     * multiple Kafka messages. The failed segment list will keep the index of
+     * of each content stream segment that had failed to be sent to Kafka, so
+     * upon retry only the failed segments are sent.
+     *
+     * @param messageContext
+     *            instance of {@link SplittableMessageContext} which hold
+     *            context information about the message to be sent
+     * @param contentStream
+     *            instance of open {@link InputStream} carrying the content of
+     *            the message(s) to be send to Kafka
+     * @param partitionKey
+     *            the value of the partition key. Only relevant is user wishes
+     *            to provide a custom partition key instead of relying on
+     *            variety of provided {@link Partitioner}(s)
+     * @return The list containing the failed segment indexes for messages that
+     *         failed to be sent to Kafka.
+     */
+    List<Integer> publish(SplittableMessageContext messageContext, InputStream contentStream, Object partitionKey) {
+        List<Integer> prevFailedSegmentIndexes = messageContext.getFailedSegments();
+        List<Integer> failedSegments = new ArrayList<>();
+        int segmentCounter = 0;
+        try (Scanner scanner = new Scanner(contentStream)) {
+            scanner.useDelimiter(messageContext.getDelimiterPattern());
+            while (scanner.hasNext()) {
+                //TODO Improve content InputStream so it's skip supported so one can zoom straight to the correct segment
+                byte[] content = scanner.next().getBytes();
+                if (content.length > 0){
+                    byte[] key = messageContext.getKeyBytes();
+                    partitionKey = partitionKey == null ? key : partitionKey;// the whole thing may still be null
+                    String topicName = messageContext.getTopicName();
+                    if (prevFailedSegmentIndexes != null) {
```

Comment:
Can we simplify this code a bit and just use:
`if (prevFailedSegmentIndexes == null || prevFailedSegmentIndexes.contains(segmentCounter)) {`
instead of replicating those 3 lines of code?